File: docs/DSA/src/py/binary_tree_maximum_path_sum.py
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def _helper(self, node):
        if not node:
            return 0

        left_max = max(self._helper(node.left), 0)
        right_max = max(self._helper(node.right), 0)
        self.res = max(self.res, node.val + left_max + right_max)
        return node.val + max(left_max, right_max)



    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.res = float('-inf')
        self._helper(root)
        return int(self.res)

File: docs/DSA/src/go/binary_tree_maximum_path_sum.go
package main

import "math"

func maxPathSum(root *TreeNode) int {
	var helper func(node *TreeNode) int
	res := math.MinInt
	helper = func(node *TreeNode) int {
		if node == nil {
			return 0
		}

		leftMax := max(helper(node.Left), 0)
		rightMax := max(helper(node.Right), 0)
		res = max(res, node.Val+leftMax+rightMax)
		return node.Val + max(leftMax, rightMax)
	}
	helper(root)
	return res
}
